src/frontends/sql/decoder_context_test.cc

```diff
@@ -127,4 +135,95 @@ TEST(DecoderContextDeathTest, DecoderContextValueRegistryDeathTest) {
                "Could not find a value with id 3.");
 }
 
+class DecodeMergeOpTest
+    : public TestWithParam<std::tuple<std::vector<Value>, std::vector<Value>>> {
+};
+
+static bool BeginsWith(absl::string_view str1, absl::string_view str2) {
+  if (str2.size() > str1.size()) return false;
+  return str1.substr(0, str2.size()) == str2;
+}
+
+static std::optional<uint64_t> ExtractIdxAfterPrefix(absl::string_view str,
+                                                     absl::string_view prefix) {
+  if (!BeginsWith(str, prefix)) return std::nullopt;
+  uint64_t result = 0;
+  std::string index_str(str.substr(prefix.size()));
+  EXPECT_NO_THROW({ result = std::stoul(index_str); });
+  return result;
+}
+
+TEST_P(DecodeMergeOpTest, DecodeMergeOpTest) {
+  auto &[direct_inputs, control_inputs] = GetParam();
+  IRContext ir_context;
+  DecoderContext decoder_context(ir_context);
+
+  const Operation &op =
+      decoder_context.MakeMergeOperation(direct_inputs, control_inputs);
+  const Block &top_block = decoder_context.BuildTopLevelBlock();
+  EXPECT_EQ(op.op().name(), DecoderContext::kSqlMergeOpName);
+  EXPECT_EQ(op.parent(), &top_block);
+  EXPECT_THAT(op.attributes(), IsEmpty());
+
+  // Reconstruct the original direct and control input vectors from the map.
+  // To do this, we start with two vectors filled entirely with dummy
+  // `Operator` values, each of which is the size of the whole map. As we
+  // traverse through the unordered map, we can then overwrite the vector
+  // at the index indicated by the key with the value from the map. We then
+  // resize each vector to the number of values inserted and compare with the
+  // original vectors.
+  Operator dummy_operator("dummy");
+  Operation dummy_operation(nullptr, dummy_operator, {}, {});
+  Value dummy_operation_result(OperationResult(dummy_operation, "dummy"));
+  const NamedValueMap &found_input_map = op.inputs();
+
+  std::vector found_direct_values(found_input_map.size(),
+                                  dummy_operation_result);
+  std::vector found_control_values(found_input_map.size(),
+                                   dummy_operation_result);
+
+  uint64_t num_direct_inputs_found = 0;
+  uint64_t num_control_inputs_found = 0;
+  for (auto &[key, value] : found_input_map) {
+    if (std::optional<uint64_t> direct_idx = ExtractIdxAfterPrefix(
+            key, DecoderContext::kMergeOpDirectInputPrefix)) {
+      found_direct_values.at(*direct_idx) = value;
```

Comment:
`found_direct_values[*direct_idx] = value`?

Reply:
`operator[]` is not bounds-checked in `std::vector`. As such, I never use it unless forced to by performance profiling.

Reply:
`operator[]` on `vector`s is not bounds checked, so I try to use `at` instead.